### ISSUE TITLE
CI fails on master: VUID-StandaloneSpirv-Flat-04744

### ISSUE THREAD

**Issue by jimblandy** [gfx-rs/wgpu#2993]
Current CI runs on `master` fail on Linux x86_64:

```
--- STDERR:                               wgpu::wgpu-tests shader_primitive_index::draw_indexed ---
WARNING: lavapipe is not a conformant vulkan implementation, testing use only.
[2022-08-27T18:09:16Z ERROR wgpu_hal::vulkan::instance] VALIDATION [UNASSIGNED-CoreValidation-Shader-InconsistentSpirv (0x6bbb14)]
    Validation Error: [ UNASSIGNED-CoreValidation-Shader-InconsistentSpirv ] Object 0: handle = 0x7f0b2483c960, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x6bbb14 | SPIR-V module not valid: [VUID-StandaloneSpirv-Flat-04744] Fragment OpEntryPoint operand 12 with Input interfaces with integer or float type must have a Flat decoration for Entry Point id 17.
      %index = OpVariable %_ptr_Input_uint Input
    
[2022-08-27T18:09:16Z ERROR wgpu_hal::vulkan::instance] objects: (type: DEVICE, hndl: 0x7f0b2483c960, name: ?)
thread 'shader_primitive_index::draw_indexed' panicked at 'UNEXPECTED TEST FAILURE DUE TO VALIDATION ERROR', wgpu/tests/common/mod.rs:307:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/panicking.rs:498:5
   1: core::panicking::panic_fmt
             at /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/panicking.rs:116:14
   2: wgpu_tests::common::initialize_test
             at ./tests/common/mod.rs:307:9
   3: wgpu_tests::shader_primitive_index::draw_indexed
             at ./tests/shader_primitive_index/mod.rs:82:5
...
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

The Vulkan spec says:

> - `VUID-StandaloneSpirv-Flat-04744`
>   Any variable with integer or double-precision floating-point type and with Input storage class in a fragment shader, must be decorated Flat

This is a recent change. [This](https://github.com/gfx-rs/wgpu/runs/7932708113?check_suite_focus=true) run passed on 8-20, while [This](https://github.com/gfx-rs/wgpu/runs/7932708113?check_suite_focus=true) run on the same sources failed on 8-27.

**Comment by jimblandy** [comment]
I didn't see anything really impressive in a comparison of the logs between the two runs, although these packages are no longer being installed:

```
libdrm-amdgpu1
libdrm-common
libdrm2
libgbm-dev
libgbm1
libglapi-mesa
libglx-mesa0
libllvm14
```

**Comment by jimblandy** [comment]
Original title and description reflected problems in the `water` example, which also fails now. Updated to match the problems in the error transcript actually shown.

**Comment by jimblandy** [comment]
Indeed, with a recent `spirv-val`:

```
$ cd wgpu/wgpu/tests/shader_primitive_index
$ naga primitive_index.wgsl primitive_index.spv
$ spirv-val --target-env vulkan1.3 primitive_index.spv
error: line 29: [VUID-StandaloneSpirv-Flat-04744] Fragment OpEntryPoint operand 27 with Input interfaces with integer or float type must have a Flat decoration for Entry Point id 31.
  %gl_PrimitiveID = OpVariable %_ptr_Input_uint Input

$ spirv-val --version
SPIRV-Tools v2022.4-dev v2022.2-85-ga98f05d0
Targets:
  SPIR-V 1.0
  SPIR-V 1.1
  SPIR-V 1.2
  SPIR-V 1.3
  SPIR-V 1.4
  SPIR-V 1.5
  SPIR-V 1.6
  SPIR-V 1.2 (under OpenCL 2.2 Full Profile semantics)
  SPIR-V 1.0 (under Vulkan 1.0 semantics)
  SPIR-V 1.3 (under Vulkan 1.1 semantics)
 SPIR-V 1.4 (under Vulkan 1.1 semantics)
 SPIR-V 1.5 (under Vulkan 1.2 semantics)
 SPIR-V 1.6 (under Vulkan 1.3 semantics)
$ 
```

**Comment by jimblandy** [comment]
This is caused by gfx-rs/naga#2032.

**Comment by nical** [comment]
Fixes in https://github.com/gfx-rs/naga/pull/2038 and https://github.com/gfx-rs/naga/pull/2035

**Comment by jimblandy** [comment]
Those Naga changes are now merged. #3008 brings them into wgpu.

**Comment by jimblandy** [comment]
This was fixed in 58bc205f.
